### PR TITLE
Add alerts UI to the map

### PR DESCRIPTION
**Pull request by lazarinibruno**
# Add alerts UI to the map

## Description
This PR introduces a bottom sheet for the map. It appears whenever the user clicks on an alert.

- If the alert was posted by the user (my alert), the buttons to edit and resolve it appear.
- If the alert was posted by another user (pal alert), the button to accept appears.

For the moment, the only button with functionality is the edit alert one, which navigates to the edit alert screen.

Moreover, this PR introduces the filtering implemented by @agonzalez-r in her PR #290.

This PR therefore closes issue #292 and #308.

## Notes
- Please feel free to play around with the map and the filtering to get a feel of it, and possibly find improvements.
- ~There is a minor bug: for the pal alerts to show up, you need to click on another screen and then return to the map. @agonzalez-r you mentioned that you know a possible solution, no?~
- **Update**: it now works :).
- While trying to debug another bug, I noticed that the alerts were constantly being refetched. A fix I found was wrapping it in a LaunchedEffect:

```kotlin
LaunchedEffect(Unit) {
  alertViewModel.fetchAlerts(
    onSuccess = {
      Log.d(TAG, "Successfully fetched alerts")
    },
    onFailure = { e -> Log.d(TAG, "Error fetching alerts: $e") },
  )
}
```

I took the liberty to modify it also in the `AlertListsScreen` (I obviously checked that everything still works and that the tests pass).

### Added
- `app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt`
- `app/src/main/res/drawable/marker_blue.xml`

### Modified
See files section.

### Removed
None.

## Dependencies added
None.

## Testing

As we already know, testing the map is complicated, since there is no direct way of interacting with it nor with its contents. Therefore, the testing that can be done is quite limited and, as a consequence, the only testing I could do is:

- Check that the filter FAB is correctly displayed and that clicking on it displays the `FilterDialog`.
- I check that clicking on the Save button on the filter calls the corresponding filtering functions. 
- I created a dedicated file `MapComponentsTest.kt` which directly tests the `MapBottomSheet`. This sheet appears in the map only after clicking on a marker, and it is thus not possible to test it through the map.

### CODE REVIEW

**Review by agonzalez-r — COMMENTED** [review]
This review is a WIP. I'll wait until you have the new changes of main. 
~I can propose a solution to figure out the bug about the alerts not showing, but for that we need to add logs, to properly figure out what's going on. I don't think we can leave it like this. 
Also, it would be great to wait until the product and urgency filter PR is merged, to make sure filtering works on the map too.~

EDIT: merging main into the branch solved the delay issues.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+ *
+ * @param alert Alert whose info will be displayed
+ */
+@Composable
+private fun AlertInfo(alert: Alert) {
+  Column {
+    Row(
+        horizontalArrangement = Arrangement.spacedBy(MaterialTheme.dimens.small3),
+        verticalAlignment = Alignment.CenterVertically,
+        modifier = Modifier.fillMaxWidth()) {
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:126]: Here basically use the structure of `MyAlertItem` and `PalsAlertItem`. I'd suggest :
- For the components `AlertProfilePicture`, `AlertTimeAndLocation`, `AlertProductAndUrgency`, `AlertAcceptButtons`, `AlertActionButton` :
  - We either move them to `AlertComponents.k` (to be discussed with @agonzalez-r since she's currently working on these components too).
  - Or we keep them in `AlertList.kt` for now (until Ale will have finished implementing on her side) but we put the above mentioned components in public, so you can call them and restructure them as you need for the map bottom sheet :)
- For the structure of `AlertInfo` we should have the same one (more or less) as the ones of `MyAlertItem` and `PalsAlertItem` which use `Modifier....weight(1f)` for good layout. The structures of the two mentioned components have weight, padding, fill / wrap the size / height / ... to enhance the UI layout on all screen sizes.
  If you still have troubles with UI design, you can contact me todayt after 14h35 (my SHS exam and we can discuss about it more + me helping you with this more technical UI design).

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+              modifier = Modifier.testTag(ALERT_URGENCY_ICON),
+          )
+        }
+    OutlinedCard {
+      Text(
+          text = alert.message,
+          modifier = Modifier.padding(MaterialTheme.dimens.small2).testTag(ALERT_MESSAGE),
+          style = MaterialTheme.typography.bodyMedium,
+      )
+    }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:192]: Honestly, outlined card isn't that pretty. We can discuss about it after my exam today.  
But more generally, we already have a "way" to display the message in `PalsAlertItem` (with a boolean which we can ignore, i.e. delete).

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+ *
+ * @param content Determines which buttons are displayed
+ * @param alert Alert whose info will be displayed
+ * @param alertViewModel Manages the alert data
+ * @param navigationActions Manages the app navigation
+ */
+@Composable
+private fun InteractionButtons(
+    content: CONTENT,
+    alert: Alert,
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:209]: Hmm, would be consistent with our code.  
However, since this function is making the difference between a `MyAlert` and a `PalsAlert`, we could directly implement the on click actions in this component. We will implement the resolve navigation to the chat in the channel task with @taghizadlaura.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            onClick = {
+              // TODO Implement alert resolution
+            },
+            modifier = Modifier.wrapContentSize().testTag(RESOLVE_ALERT_BUTTON),
+            colors = getFilledPrimaryContainerButtonColors(),
+        ) {
+          Icon(imageVector = Icons.Outlined.Check, contentDescription = "Resolve alert")
+          Spacer(modifier = Modifier.width(MaterialTheme.dimens.small2))
+          Text(text = RESOLVE_BUTTON_TEXT, style = MaterialTheme.typography.bodyLarge)
+        }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:247]: Doesn't make much sense to have an `Edit` and a `Resolve` button since they both lead us to the edit alert screen. I'd suggest to just keep the edit one.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+          Icon(imageVector = Icons.Outlined.Check, contentDescription = "Accept alert")
+          Spacer(modifier = Modifier.width(MaterialTheme.dimens.small2))
+          Text(text = ACCEPT_BUTTON_TEXT, style = MaterialTheme.typography.bodyLarge)
+        }
+      }
+    }
+  }
+}
+
+fun trimLocationText(locationText: String): String {
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:268]: - Hehe I like the "..."
- Should be in `AlertComponents.kt` since we're going to use it quite a bit.
- Also documentation as @agonzalez-r said.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            onClick = {
+              // TODO Implement alert accept
+            },
+            modifier = Modifier.wrapContentSize().testTag(ACCEPT_ALERT_BUTTON),
+            colors = getFilledPrimaryContainerButtonColors(),
+        ) {
+          Icon(imageVector = Icons.Outlined.Check, contentDescription = "Accept alert")
+          Spacer(modifier = Modifier.width(MaterialTheme.dimens.small2))
+          Text(text = ACCEPT_BUTTON_TEXT, style = MaterialTheme.typography.bodyLarge)
+        }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:262]: - Update the documentation from `// Edit` to `// Resolve`.
- As mentioned earlier in another comment, `InteractionButtons` doesn't necessary have to have the `onClick` parameter since it has to differentiate between my alert and pal's alert.
  However, we will implement the navigation to the channel later on in channel task as mentioned earlier.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 7]
```diff
@@ -85,6 +116,78 @@ fun MapScreen(
...
+    }
+  }
+
+  val onPalAlertClick: (Alert) -> Unit = remember {
+    { alert ->
+      showBottomSheet = true
+      content = CONTENT.PAL_ALERT
+      alertViewModel.selectAlert(alert)
+    }
+  }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:138]: - We don't tend to extract the `onClick` actions into `val`s since it makes the code harder to understand (have to scroll all the way up to see the on click action). Please remove them and use them "raw" in the code.
- Plus they're only used once, not the purpose of such a `val`.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 8]
```diff
@@ -85,6 +116,78 @@ fun MapScreen(
...
+
+  // To manage the alert filter state
+  var isFilterApplied by remember { mutableStateOf(false) }
+  var showFilterDialog by remember { mutableStateOf(false) }
+  var selectedLocation by remember { mutableStateOf<Location?>(null) }
+  var radiusInMeters by remember { mutableDoubleStateOf(DEFAULT_RADIUS) }
+  var productFilter by remember { mutableStateOf<Product?>(Product.NO_PREFERENCE) }
+  var urgencyFilter by remember { mutableStateOf<Urgency?>(null) }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:146]: - Good.
- ```suggestion
    // To manage the alert filter state
    var isFilterApplied by remember { mutableStateOf(false) }
    var showFilterDialog by remember { mutableStateOf(false) }
    var selectedLocation by remember { mutableStateOf<Location?>(null) }
    var radiusInMeters by remember { mutableDoubleStateOf(DEFAULT_RADIUS) }
    var productFilter by remember { mutableStateOf<Product?>(null) }
    var urgencyFilter by remember { mutableStateOf<Urgency?>(null) }
  ```
  Same in `AlertLists`.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 9]
```diff
@@ -85,6 +116,78 @@ fun MapScreen(
   val myLocationOverlay = remember { FolderOverlay() }
   val alertOverlay = remember { FolderOverlay() }
 
+  // To manage the bottom sheet state
+  val sheetState = rememberModalBottomSheetState()
+  val scope = rememberCoroutineScope()
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:121]: Is it really good practice to extract `rememberCoroutineScope()` here ? Especially if we're using it only once ? I feel like it's harder to understand the code by extracting it here.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: I noticed a few bugs :
- When I filter locations using the radius, only the pals' alerts are filtered. I think it doesn't bother too much since one still sees my alerts regardless of the radius. I don't know in which context one would like to filter his own alerts.
- I can enter an invalid location, apply filter. The filter will display as active although the filter is not filtering with a valid location.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 10]
```diff
@@ -93,101 +196,135 @@ fun MapScreen(
...
+              onClick = { recenterOnMyLocation(mapView, myLocation) },
+              modifier = Modifier.testTag(C.Tag.MapScreen.MY_LOCATION_BUTTON),
+          ) {
+            Icon(
+                imageVector = Icons.Outlined.MyLocation,
+                contentDescription = "Recenter on my position",
+            )
+          }
+
+          Spacer(modifier = Modifier.height(MaterialTheme.dimens.small3))
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:226]: Use `spacedBy` instead of a Spacer :
```kotlin
        Column(
            modifier = Modifier.wrapContentSize(),
            verticalArrangement =
                Arrangement.spacedBy(MaterialTheme.dimens.small2, Alignment.CenterVertically),
        ) {
          FloatingActionButton(...
```

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 11]
```diff
@@ -205,51 +342,77 @@ private fun initializeMap(
...
 private fun updateAlertMarkers(
     mapView: MapView,
     alertOverlay: FolderOverlay,
     context: Context,
-    alertList: List<Alert>
+    myAlertsList: List<Alert>,
+    palAlertsList: List<Alert>,
+    onMyAlertClick: (Alert) -> Unit,
+    onPalAlertClick: (Alert) -> Unit
 ) {
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:358]: Update documentation.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 12]
```diff
@@ -93,101 +196,135 @@ fun MapScreen(
...
+                imageVector = Icons.Outlined.MyLocation,
+                contentDescription = "Recenter on my position",
+            )
+          }
+
+          Spacer(modifier = Modifier.height(MaterialTheme.dimens.small3))
+
+          // Filter button
+          FilterFab(isFilterApplied) { showFilterDialog = true }
+        }
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:230]: Instead of using the spacer, we can use
```kotlin
        Column(
            modifier = Modifier.wrapContentSize(),
            verticalArrangement =
                Arrangement.spacedBy(MaterialTheme.dimens.small2, Alignment.CenterVertically),
        ) {
          // Recenter button
          FloatingActionButton(...

```

File: app/src/test/java/com/android/periodpals/ui/map/MapComponentsTest.kt
Comment (francelu) [inline app/src/test/java/com/android/periodpals/ui/map/MapComponentsTest.kt:1]: Test pass ! :)

**Review by francelu — CHANGES_REQUESTED** [review]
## Summary
This PR introduces several significant updates to the PeriodPals app. It primarily focuses on enhancements to the MapScreen. Changes include updates to the alert markers, location management, and the addition of the filter.


## Important
### Code Quality
- The code is generally well-structured and adheres to Kotlin coding conventions.
- The formatting changes in `MapScreen` improve readability by breaking down complex functions into smaller, more manageable parts.
- The introduction of test tags in `C.kt` is helpful for maintaining consistency.

### Functionality
- The updates to `MapScreen` and the new `MapComponents` enhance the functionality by adding new features such as alert filters and bottom sheets.
- The changes ensure that alert markers are more interactive and provide better user feedback.
- The integration of `LocationViewModel` adds robustness to location management.

### Performance
- The use of `LaunchedEffect` for fetching alerts and updating markers is efficient and ensures that the UI remains responsive.
- The changes to marker updates on the map are well-optimized.

### Security
- No significant security concerns noted in the changes.


## Testing
- The added tests in `MapScreenTest` ensure that the new functionalities are well-covered.
- The tests for filter functionality and alert fetching are comprehensive and verify that the features work as intended.


## Documentation
- The PR lacks sufficient documentation for the newly introduced functions and components. Adding KDoc comments would be beneficial for future maintainability.


## Suggestions
- Ensure that all new functions, especially those in `MapComponents.kt`, are well-documented.
- Consider adding more edge case tests for the new filter functionality to cover scenarios like invalid location inputs.


## Steps before PR approved
- Add KDoc comments to all newly introduced functions and components.
- Implement additional edge case tests for the filter functionality.
- Check comments / resolve.
- Contact me if you want :) after my SHS for UI design.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: Hi, the first one is not a bug... we decided the user must always be able to see their own alerts, even if they are outside the radius. 

For the second one, I don't think that should be solved in this PR, the implementation of the filter was already done, reviewed and merged with main, so we would need to fix that separately. What exactly do you mean by invalid location? I think I only check that the location is not null iff the radius is set (moved to different than default). I must specify that the user **can apply a filter only on product/urgency**, and not location specified.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 7 above)
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:138]: I agree here

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt (hunk 1 above)
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:126]: You can extract the components in `AlertComponents`, it's okay by me !

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: Screenshot 2024-12-17 at 15 39 58

the Location field component already checks that the input in the field has to be valid. It is normal behaviour to be able to left the location empty (except if a radius is also indicated)

However I agree that we don't test that **no filters** are applied when clicking on the `Apply` button, i.e at least one filter should be applied

**Review by agonzalez-r — APPROVED** [review]
## Summary
Thank you for this PR, it effectively implements the alert view from map feature, which is really important since it is actually the main purpose of the map screen: the user is able to visualise the alerts near them, apply filters, and ultimately choose one to respond.
## Important (which includes)
#### Code Quality
- Code is clean. 
- I have some doubts about the consistency, since the way you implemented the bottom sheet component is by passing the view models as parameter, instead of passing the `onClick` function for the button like we do in the rest of the code. But it does work well in this context.
#### Functionality
- @francelu made a remark about an edge case: when clicking in "Apply filter" button when no filter was indicated, it lets the user click on the button anyways. Please fix this in `FilterDialog()` by adding a line on the `onClick` of this button: 
```
(in apply filter button...)
onClick = {
if(location != null || product != NO_PREFERENCE || urgency != null){
onSave() ...
} else{
//show toast
}
}
```
something of that logic...
#### Performance 
- Code works great, no delay in showing the markers! Also no problems with the filters.
- **The UI does need to be improve.** But I know that you are already working on that, directly with France, so I'll let you do that.  

## Testing
- LGTM 

## Documentation
- Some documentation was missing, see the comments. 

## Suggestions
- As @francelu mentioned, you could use the components from `AlertLists.kt` for the display of the alerts information, if you do so, please move them to `AlertComponents.kt`

## Steps before PR approved
- Improve UI. 
- Apply France's suggestion (except the ones I expressed my disagreement...)

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:268]: Okay, I will move it to `AlertComponents.kt` and I will add the documentation.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+/**
+ * Displays the alert info: name of user, location, time, product type and urgency.
+ *
+ * @param alert Alert whose info will be displayed
+ */
+@Composable
+private fun AlertInfo(alert: Alert) {
+  Column {
+    Row(
+        horizontalArrangement = Arrangement.spacedBy(MaterialTheme.dimens.small3),
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:124]: Mhh, good question. It would look like this:

image


Maybe with proper padding I can make it look nicer. I'll try out some stuff.

File: app/src/main/java/com/android/periodpals/ui/alert/AlertLists.kt [hunk 14]
```diff
@@ -107,6 +104,7 @@ private enum class AlertListsTab {
  * @param alertViewModel The view model for managing alert data.
  * @param authenticationViewModel The view model for managing authentication data.
  * @param navigationActions The navigation actions for handling navigation events.
+ * @param locationViewModel The view model for managing the location data.
  * @param gpsService The GPS service that provides the device's geographical coordinates.
  * @param navigationActions The navigation actions for handling navigation events.
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/alert/AlertLists.kt:109]: Sure, I'll do it. However, it wasn't me who did this doc 😭😭

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:1]: Would you rather have it in the `map/` directory?

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+ * @param content Determines which buttons are displayed
+ * @param alertViewModel Manages the alert data
+ * @param navigationActions Manages the app navigation
+ */
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun MapBottomSheet(
+    sheetState: SheetState,
+    onDismissRequest: () -> Unit,
+    onHideRequest: () -> Unit,
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:84]: Good catch!

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+  ) {
+    Column(
+        verticalArrangement = Arrangement.spacedBy(MaterialTheme.dimens.small1),
+        modifier =
+            Modifier.padding(
+                start = MaterialTheme.dimens.small3,
+                end = MaterialTheme.dimens.small3,
+                bottom = MaterialTheme.dimens.small3,
+            ),
+    ) {
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:103]: Yes, thanks for you suggestion!

- Using `modifier = Modifier.fillMaxWidth().wrapContentHeight().testTag(BOTTOM_SHEET)` is certainly a good improvement.
- The same applies for `Arrangement.spacedBy(MaterialTheme.dimens.small2, Alignment.CenterVertically)`
- However, for the padding, I prefer to leave it as is. The problem with adding padding on all 4 "edges" is that it adds unnecessary padding at the top: 

image

I think it looks better without, but if you prefer I also like with padding on top.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt (hunk 4 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:247]: Yes, I agree with @agonzalez-r. The buttons have different goals.

Imagine that you post an alert, but then you notice that you actually had tampons in your backpack. Then, you can click on resolve to remove it from the posted alerts.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 7 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:138]: Yes, in retrospective, I don't know why I did this hahah.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 8 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:146]: Oki.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 9 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:121]: In the end, I don't use `rememberCoroutineScope()` so I removed it.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 10 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:226]: Good suggestion

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 11 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:358]: Good catch!

**Comment by lazarinibruno** [comment]
> * I have some doubts about the consistency, since the way you implemented the bottom sheet component is by passing the view models as parameter, instead of passing the `onClick` function for the button like we do in the rest of the code. But it does work well in this context.

Yes, I agree with you. I think last week, when I finished implementing it, I was very tired haha. I'll change it to not pass the whole view model as parameter.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 8 above)
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt]: Actually, if we want to be consistent with the `AlertLists`, we should leave it like it was:

```kotlin
var productFilter by remember { mutableStateOf<Product?>(Product.NO_PREFERENCE) }
```

I think @agonzalez-r gave me a reason for it too, but I don't remember. It has to do with the way the filter works.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:1]: I'd say either leave it here, or move it into `ui/map/Map.kt` directly. But then, that one would be a very long file.

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt]: > However, for the padding, I prefer to leave it as is. The problem with adding padding on all 4 "edges" is that it adds unnecessary padding at the top:

Sure it's fine with just how you had it at the beginning.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt (hunk 8 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:129]: Okay if you want, why not.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: > Hi, the first one is not a bug... we decided the user must always be able to see their own alerts, even if they are outside the radius.

@agonzalez-r  
Good then !

File: app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/components/MapComponents.kt:1]: Okay, I see. I think I prefer having two files. Otherwise, the file would be too long, as you said.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 17]
```diff
@@ -205,51 +342,77 @@ private fun initializeMap(
...
+    myAlertsList: List<Alert>,
+    palAlertsList: List<Alert>,
+    onMyAlertClick: (Alert) -> Unit,
+    onPalAlertClick: (Alert) -> Unit
 ) {
   alertOverlay.items.clear()
-  alertList.forEach { alert ->
+
+  // Draw markers for my alerts
+  myAlertsList.forEach { alert ->
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:360]: Its true that they are quite similar, but they aren't the same. The title changes, the icon, and the callback too. I can see if its worth extracting it into a function, but it seems a bit too over-complicated.

[the remark this answers is not in the document]
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:354]: It's okay to leave it like that as well

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: I see. 
In my opinion the real bug here is that it lets the user click on "Apply Filter" even though no filter was indicated. After all, entering an invalid location a don't clicking on any of the option of the dropdown menu, is simply not selecting a location. 
To fix this bug I propose this solution: 
```
onClick = {
              if ((sliderPosition != 100f) &&
                  (location == null)) { // if the user selects a radius but not a location
                Toast.makeText(context, ERROR_MESSAGE_INVALID_LOCATION, Toast.LENGTH_SHORT).show()
              } else if (location == null &&
                  product == FILTERS_NO_PREFERENCE_TEXT &&
                  urgency == FILTERS_NO_PREFERENCE_TEXT) { // if the user does not select any filter
                Toast.makeText(context, "Please select at least one filter", Toast.LENGTH_SHORT)
                    .show()
              } else {
                onSave(sliderPosition.toDouble(), selectedProduct, selectedUrgency)
                onDismiss()
              }
            }
```
The new part is the `else if`, which checks if at least one filter was selected. 

I can commit these changes if you agree @lazarinibruno, or you can do it yourself.

Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: Okay, in the end, I opted for the following solution since it was the simplest:

```kotlin
isFilterApplied =
    (radius != 100.0) || (productFilter != Product.NO_PREFERENCE) || (urgencyFilter != null)
```

So, if the user has no preference for anything, after clicking on apply filter, the bubble wont appear since no alerts are being filtered.

Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:1]: Okay, I will modify it too then.

**Comment by lazarinibruno** [comment]
The code duplication in this PR is quite high (8.1%). This is because the map and the alert list both use the `FilterDialog`, which has a lot of parameters among which many of them are callbacks which affect local variables of the composables calling the dialog.

Therefore, it is not possible to extract the code and make it be shared between the map and the alert list.

**Review by agonzalez-r — APPROVED** [review]
LGTM 
You effectively fixed the bugs we encountered, which were edge cases but still important. 
About the duplication percentage, it could be fixed, but it's not that bad and we really need to get this code merged.

**Review by francelu — APPROVED** [review]
LGTM. Code duplication is due to some similarities between Map and AlertLists of which both require some vals to remember the filter states.
